Accessibility: fixing BoundingRectangle property for PropertyGrid entries

vladimir-krestov opened the pull request:
fixes #856

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=857) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=857) before we can accept your contribution.<br/><br/>:x: vladimir-krestov [sign now](https://cla.dotnetfoundation.org/dotnet/winforms?pullRequest=857)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/winforms?pullRequest=857) it.</sub>

vladimir-krestov commented:
Done: 
- Fixed implementation of PropertyGridView entries accessible rectangle bounds getting
- Added unit test to check PropertyGridView entries height

In plans:
- Fix getting Offscreen property for invisible PropertyGrid entries

vladimir-krestov commented:
Done:
- Fixed getting Offscreen property for invisible PropertyGrid entries

danmosemsft commented on src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs:
```diff
@@ -784,9 +787,21 @@ public PropertyGridView(IServiceProvider serviceProvider, PropertyGrid propertyG
 
+            Rectangle parent = gridEntry.OwnerGrid.GridViewAccessibleObject.Bounds;
+
+            if (pt.y > parent.Bottom - 1) // - 1 this is PropertyGridView bottom border 
```
Seems extracting `int propertyGridViewBottomBorder = parent.Bottom -1;` might be worthwhile.

vladimir-krestov replied:
Fixed it. Thanks

vladimir-krestov commented:
@zsd4yr, the unit test is already included in the pull request.
It is here: src / System.Windows.Forms / tests / UnitTests / AccessibleObjects / PropertyGridViewRowsAccessibleObjectTests.cs. 
Let me know if more tests are needed.

zsd4yr commented:
@vladimir-krestov looks good to me :) thanks for the tests! You should have merge permissions, so please merge at your pleasure!

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs:
```diff
@@ -772,9 +772,12 @@ public PropertyGridView(IServiceProvider serviceProvider, PropertyGrid propertyG
             int row = GetRowFromGridEntry(gridEntry);
-            if (row == -1) {
+
+            if (row < 0)
```
When do we get negative row indices other than -1?

vladimir-krestov replied:
```suggestion
            int row = GetRowFromGridEntry(gridEntry);
```
This method returns -1 in case entry == null. If the current entity is higher than the first visible entity, then this method returns the negative index. This way I fixed the incorrect display of bounds above the visible area.

Tanya-Solyanik replied:
Thank you for the explanation

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/PropertyGridView.cs:
```diff
@@ -784,9 +787,23 @@ public PropertyGridView(IServiceProvider serviceProvider, PropertyGrid propertyG
+
+            if (pt.y > propertyGridViewBottom)
+            {
+                return new Rectangle(0, 0, 0, 0);
```
Rectangle.Empty?

vladimir-krestov replied:
Thank you. Fixed.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewRowsAccessibleObjectTests.cs:
```diff
@@ -0,0 +1,93 @@
+{
+    public class PropertyGridViewRowsAccessibleObjectTests
+    {
+        [Fact]
```
I wonder that since you do Application.Run and start message loop, perhaps this should be a STAFact?

vladimir-krestov replied:
I changed attribute to [StaFact]. Thanks.
I added the unit test with Application.Run because of entries rectangle bounds don't have time to initialize and entries bounds height is zero.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/PropertyGridViewRowsAccessibleObjectTests.cs:
```diff
@@ -0,0 +1,93 @@
+            this.ClientSize = new Size(508, 367);
+            this.Controls.Add(propertyGrid);
+            this.Controls.Add(domainUpDown);
+            this.Load += new EventHandler(this.TestForm_Load);
```
```suggestion
            Load += TestForm_Load;
```

vladimir-krestov replied:
Thank you. Fixed.

weltkante commented:
> the test is running in an MTA

Does WinForms even run reliably on MTA? Most WinForm handles are bound to a threads message queue and don't care about apartments, accessing them from a different thread is not part of the contract of most native APIs.

Also there are various integral parts of WinForms which need to access COM and if those initializations happen on the wrong apartment you are in big trouble because COM objects remember the apartment and try to get back on it. If you have bad luck the GC gets deadlocked by a COM objects finalizer trying to get back on a thread which is busy doing other stuff. Took us several months to figure that one out.

If you didn't do already you should probably invest in proper STA handling of your test infrastructure.

merriemcgaw commented:
@vladimir-krestov once @Tanya-Solyanik 's comments are addressed we'll go ahead and merge this.
